### ISSUE TITLE
feat: TypeScript support

### ISSUE THREAD

**Issue by itskibo** [nexxtway/react-rainbow#942]
-   [x] I have searched the [issues](https://github.com/nexxtway/react-rainbow/issues) of this repository and believe that this is not a duplicate.

**Is your feature request related to a problem? Please describe.**
TypeScript is in more demand these days, and rapidly being adopted. Especially in professional environments, type safety is a large issue that's being thought of more these days (e.g. PHP looking at it with 7.x).
This library seems interesting to me, and I want to ensure that I can properly use it with personal projects.

## Expected Behavior 🤔
Correct typings when importing components.

## Context 🔦
I feel like the context is pretty clear: type safety for those who want it.

One could argue to add the definitions to DefinitelyTyped, but then we're maintaining both this repository AND the types over at DT.

**Describe the solution you'd like**
TypeScript support.

**Describe alternatives you've considered**
None.

**Comment by gustavopch** [comment]
TS support is a make-or-break deal for a lot of projects.

**Comment by rpmonteiro** [comment]
Exactly my situation... this library looks amazing, but until TS support is here, it can't be used.

**Comment by gustavopch** [comment]
Refactoring to TS should probably also come with:
- Use named modules wherever possible (see https://basarat.gitbooks.io/typescript/docs/tips/defaultIsBad.html).
- Use functional components and Hooks wherever possible as they are easier to type (not to mention a bunch of other benefits).

**Comment by reiniergs** [comment]
add TS is a high priority now since I will increase adoption :)

**Comment by gustavopch** [comment]
> add TS is a high priority now since I will increase adoption :)

Would you accept a PR to refactor the whole code base to TS? I may give it a try.

**Comment by reiniergs** [comment]
that's the whole purpose of OSS right? please go for it!

**Comment by gustavopch** [comment]
@reiniergs Great! Just a problem: I cloned the repo and ran `yarn test`, but I'm getting a lot (hundreds I guess) of failed tests. Is that right?

**Comment by reiniergs** [comment]
CI is passing, Jest test should be all passing locally ... I have no idea.

**Comment by gustavopch** [comment]
Installed Node.js v11.8.0 to match the CI version, but the tests still fail. Mostly related to prop types, like: ``Warning: Failed prop type: The prop `label` is marked as required in `Lookup`, but its value is `undefined`.``
No idea what to do. I'm using Arch Linux BTW.

**Comment by LeandroTorresSicilia** [comment]
@gustavopch 
the warnings showed in console are due to bad use of proptypes in tests but this is not related to failing tests
if you can show me the log output of running tests

**Comment by reiniergs** [comment]
```
Test Suites: 254 passed, 254 total
Tests:       1369 passed, 1369 total
Snapshots:   1 passed, 1 total
Time:        21.871s
Ran all test suites matching /.\/src/i.
✨  Done in 24.49s.
```
tests are passing for me locally

**Comment by gustavopch** [comment]
@LeandroTorresSicilia Here: https://pastebin.com/1qX19x6h

**Comment by LeandroTorresSicilia** [comment]
@gustavopch 
I see the log, there is a bunch of warning but there is not log saying that the tests failed

**Comment by gustavopch** [comment]
@LeandroTorresSicilia Oh, forget it. Those are just warnings. The tests are indeed passing. :smile:

**Comment by stephen-zhao** [comment]
I was just about to jump on to use this library when first I saw the post over at hacker news.. glad to see TS support coming in because I can't use it without typings. Can't wait to see the PR.

**Comment by LeandroTorresSicilia** [comment]
@stephen-zhao 
Adding TS support is right now high priority and it should be as fast as possible, but it is not as simply as rewrite

**Comment by itskibo** [comment]
I don't think it requires a (large) rewrite, at least not for now. Adding typings should be sufficient as a first step.

**Comment by gustavopch** [comment]
@LeandroTorresSicilia I won't be able to submit a PR these days. Got busy with work here. :disappointed:

**Comment by stephen-zhao** [comment]
In that case, we can add typings incrementally, like one PR per component, just to break it down into more manageable units. I can try my hand at a couple..

**Comment by miguelramos** [comment]
Notice that current development environment doesn't support compile in typescript. Plus a huge list of branches exists here in remote repo and just a thought wouldn't be more clean and structured if adopt lerna/monorepo style in here to help development environment? And them maybe typescript comes with that... Just asking if you guys are open to that.

**Comment by LeandroTorresSicilia** [comment]
> wouldn't be more clean and structured if adopt lerna/monorepo style in here to help development environment?
Yes, we already talk about use lerna since we have other projects related to this and plan to add more, that is in our roadmap

**Comment by miguelramos** [comment]
> > wouldn't be more clean and structured if adopt lerna/monorepo style in here to help development environment?
> 
> Yes, we already talked about use lerna since we have other projects related to this and plan to add more, it is in our roadmap

Nice. So are you open to a PR on this? Does my above propose help? Na would you accept help to start this?

**Comment by stephen-zhao** [comment]
I was trying to set up for typescript when I ran into an issue. I had to upgrade to Babel ^7.0.0 which has new package names. I can make a PR for this upgrade if that's fine.

**Comment by LeandroTorresSicilia** [comment]
Hey guys the new version 1.7.0 has fully support of typescript

**Comment by LeandroTorresSicilia** [comment]
fixed by #1012
